src/Microsoft.ML.Tokenizers/Tokenizer.cs

```diff
@@ -35,7 +35,23 @@ public abstract class Tokenizer
         /// <param name="textSpan">The span of the text to encode which will be used if the <paramref name="text"/> is <see langword="null"/>.</param>
         /// <param name="settings">The settings used to encode the text.</param>
         /// <returns>The encoded results containing the list of encoded Ids.</returns>
-        protected abstract EncodeResults<int> EncodeToIds(string? text, ReadOnlySpan<char> textSpan, EncodeSettings settings);
+        protected virtual EncodeResults<int> EncodeToIds(string? text, ReadOnlySpan<char> textSpan, EncodeSettings settings)
```

Comment:
Before this change, we didn't have to do it because there was no `EncodeToTokens` need the max token count. But now we do.

Reply:
There was in that the user requested it. They could be passing in a giant document and only want the first 100 tokens. If the implementation doesn't respect max token count, it'll end up doing way more work and allocating way more memory than the user asked it to. I could imagine there might also be functional ramifications to not respecting it, e.g. if there's some state later in the input that couldn't be tokenized, such that an exception would occur, and max token count is used by the caller to avoid getting to that point.

Regardless of this change, if there's a setting like "max tokens to process" being passed as input to a method, I think it's a bug if the method ignores it and produces results for more than that.

Reply:
The method EncodeToTokens which takes `EncodeSettings` is a `protected` which I think users will not be able to call it directly. They can only call 

```C#
public IReadOnlyList<EncodedToken> EncodeToTokens(string text, out string? normalizedString, bool considerPreTokenization = true, bool considerNormalization = true)
public IReadOnlyList<EncodedToken> EncodeToTokens(ReadOnlySpan<char> text, out string? normalizedString, bool considerPreTokenization = true, bool considerNormalization = true) 
```

These are not taking max count so there are no today users can pass max token counts. 

Anyway, I think we agree that we need to add EncodeToToken doc explaining the expectation.

Reply:
Thanks.

What's the rationale for:
```
public IReadOnlyList<EncodedToken> EncodeToTokens(string text, out string? normalizedString, bool considerPreTokenization = true, bool considerNormalization = true)
```
taking considerPreTokenization and considerNormalization directly rather than taking an EncodeSettings?

Reply:
In the design review, `EncodeSettings` is introduced to reduce the number of methods that the user needs to override. The original proposal had the following methods as abstract/virtual. The idea was to have one method to override which needed to introduce `EncodeSettings`. 

```C#
public IReadOnlyList<EncodedToken> EncodeToTokens(string text, out string? normalizedString, bool considerPreTokenization = true, bool considerNormalization = true)
public IReadOnlyList<EncodedToken> EncodeToTokens(ReadOnlySpan<char> text, out string? normalizedString, bool considerPreTokenization = true, bool considerNormalization = true) 
```

And then we applied the same idea to other APIs (e.g. EncodeToIds) which is required to allow max token count and we needed to add that to `EncodeSettings` too. The intention was not exposing directly APIs taking `EncodeSettings`. I know nothing prevents us from exposing it if we think it is useful.